cluster-autoscaler/core/scaledown/actuation/delete_in_batch_test.go

```diff
@@ -0,0 +1,169 @@
+		},
+		{
+			name:        "Add nodes that belong to one nodeGroup",
+			nodes:       nodes1,
+			wantBatches: 1,
+		},
+		{
+			name:        "Add 3 nodes that belong to 2 nodeGroups",
+			nodes:       []*apiv1.Node{nodes1[0], nodes2[0], nodes2[1]},
+			wantBatches: 2,
+		},
+		{
+			name:        "Add 3 nodes that belong to 2 nodeGroups, all nodes are drained",
+			nodes:       []*apiv1.Node{nodes1[0], nodes2[0], nodes2[1]},
+			wantBatches: 2,
+			drained:     true,
+		},
+	}
+	for _, test := range testcases {
+		d := NodeBatcherDeleterImpl{
+			ctx:                   &ctx,
+			clusterState:          nil,
+			nodeDeletionTracker:   nil,
+			deletionsPerNodeGroup: make(map[string][]*apiv1.Node),
+			drainedNodeDeletions:  make(map[string]bool),
+		}
+		batchNames := 0
+		for _, node := range test.nodes {
+			_, first, err := d.addNodeToBucket(node, test.drained)
+			if err != nil {
+				t.Errorf("addNodeToBucket return error %q when addidng node %v", err, node)
+			}
+			if first {
+				batchNames += 1
+			}
+		}
+		if batchNames != test.wantBatches {
+			t.Errorf("Want %d batches, got %d batches", test.wantBatches, batchNames)
+		}
+
+	}
+}
+
+func TestCleanUp(t *testing.T) {
+	provider := testprovider.NewTestCloudProvider(nil, func(id, nodeName string) error {
+		return nil
+	})
+	fakeClient := &fake.Clientset{}
+	fakeLogRecorder, _ := clusterstate_utils.NewStatusMapRecorder(fakeClient, "kube-system", kube_record.NewFakeRecorder(5), false, "my-cool-configmap")
+	ctx, err := NewScaleTestAutoscalingContext(config.AutoscalingOptions{}, nil, nil, provider, nil, nil)
+
+	clusterStateRegistry := clusterstate.NewClusterStateRegistry(provider, clusterstate.ClusterStateRegistryConfig{}, fakeLogRecorder, NewBackoff())
+	if err != nil {
+		t.Fatalf("Couldn't set up autoscaling context: %v", err)
+	}
+	nodeGroup1 := "ng-1"
+	nodeGroup2 := "ng-2"
+	nodesDrained := generateNodes(5, "ng-1")
+	nodesEmpty := generateNodes(5, "ng-2")
+	provider.AddNodeGroup(nodeGroup1, 1, 10, 5)
+	provider.AddNodeGroup(nodeGroup2, 1, 10, 5)
+	for _, node := range nodesDrained {
+		provider.AddNode(nodeGroup1, node)
+	}
+	for _, node := range nodesEmpty {
+		provider.AddNode(nodeGroup2, node)
+	}
+	d := NodeBatcherDeleterImpl{
+		ctx:                   &ctx,
+		clusterState:          clusterStateRegistry,
+		nodeDeletionTracker:   deletiontracker.NewNodeDeletionTracker(1 * time.Hour),
+		deletionsPerNodeGroup: make(map[string][]*apiv1.Node),
+		drainedNodeDeletions:  make(map[string]bool),
+	}
+	for _, node := range nodesDrained {
+		_, _, err := d.addNodeToBucket(node, true)
+		if err != nil {
+			t.Errorf("addNodeToBucket return error %q when addidng node %v", err, node)
+		}
+	}
+	for _, node := range nodesEmpty {
+		_, _, err := d.addNodeToBucket(node, false)
+		if err != nil {
+			t.Errorf("addNodeToBucket return error %q when addidng node %v", err, node)
+		}
+	}
+	batches := len(d.deletionsPerNodeGroup)
+
+	// CleanUp NodeGroup that is not present in bucket.
```

Comment:
Why not make this a table-style test as well? I'd like to see a test case where we do `addNode, addNode, remove, addNode, remove for the same node group in quick succession (or maybe all but the last remove started at the same time as goroutines, and then a synchronous remove)`. Then assert that all nodes get deleted, and that the state is as expected at the end (empty `deletionsPerNodeGroup`, empty `drainedNodeDeletions`). I know it should work with the current implementation, but I can see somebody changing something in the future that would make us skip some node in this scenario, or leak memory in the structures.

Reply:
The test now is table-style. This is unit test, that verify remove() call only. The whole batch, i.e AddNode() should be tested in actuator_test, that test the whole functionality.